node/actors/bft/src/leader/state_machine.rs

```diff
@@ -1,17 +1,19 @@
-use crate::{metrics, ConsensusInner};
+use crate::{metrics, ConsensusInner, PayloadSource};
 use std::{
     collections::{BTreeMap, HashMap},
+    sync::Arc,
     unreachable,
 };
 use tracing::instrument;
-use zksync_concurrency::{ctx, metrics::LatencyHistogramExt as _, time};
+use zksync_concurrency::{ctx, error::Wrap as _, metrics::LatencyHistogramExt as _, time};
 use zksync_consensus_roles::validator;
 
 /// The StateMachine struct contains the state of the leader. This is a simple state machine. We just store
 /// replica messages and produce leader messages (including proposing blocks) when we reach the threshold for
 /// those messages. When participating in consensus we are not the leader most of the time.
-#[derive(Debug)]
 pub(crate) struct StateMachine {
+    /// Payload provider for the new blocks.
+    pub(crate) payload_source: Arc<dyn PayloadSource>,
```

Comment:
Dumb question: Is using `Arc` instead of `Box` and `&self` receiver on `propose` instead of `&mut self` motivated by some external reasoning? Or is it just chosen defensively? Within the repo, both can be changed without a problem. IMO, unlike storage, there's no inherent expectation that a proposer is shared.

Reply:
Well, ptal https://github.com/matter-labs/zksync-era/pull/554 . Here storage is payload source as well. I guess I can make it `Box<Arc<Storage>>` there, if you want.

Reply:
Sorry, I've looked at this PR first. No need to box an `Arc`, although you could theoretically split off `PayloadSource` implementation to a different struct than the one implementing `BlockStore` (it'd essentially wrap a `ConnectionPool`).

Reply:
ack